feat(replays): added heap memory chart to replay details

Danondso opened the pull request:
# Description
Closes #33623 
Added area chart of heap memory usage over the duration of the session.

# Screenshots
<img width="1037" alt="image" src="https://user-images.githubusercontent.com/7014871/165967010-bc7509be-396d-404c-9129-170c2f1f25fc.png">

Full screen light theme:
<img width="1724" alt="image" src="https://user-images.githubusercontent.com/7014871/165967161-ec94836e-a742-4721-b75c-6245f18e6538.png">

When there's no memory events: 
<img width="1724" alt="image" src="https://user-images.githubusercontent.com/7014871/165967238-7dfa5c47-f749-4a11-96cf-4fcea62f00b9.png">



<!-- Describe your PR here. -->



<!--

  The following applies to third-party contributors.
  Sentry employees and contractors can delete or ignore.

-->

----

By submitting this pull request, I confirm that Sentry can use, modify, copy, and redistribute this contribution, under Sentry's choice of terms.

Danondso committed:
feat: added a memory tab on the new replays details ui
feat: added an area chart for heap memory usage

Danondso committed:
feat: added fallback message if no memory spans exist.

Danondso committed:
Merge branch 'master' into feat/replays/33623/heap-memory-chart

Danondso commented on static/app/views/replays/detail/focusArea.tsx:
```diff
@@ -29,11 +30,21 @@ function FocusArea(props: Props) {
   const {routes, router} = useRouteContext();
   const organization = useOrganization();
-
+  // memoize the memory events coming from the event because we reassign
+  // the data without the memory events within the 'performance' case
+  const memorySpans = useMemo(
+    () => eventWithSpans?.entries[0]?.data?.filter(datum => datum?.data?.memory) || [],
+    [eventWithSpans?.entries[0]]
+  );
   switch (active) {
     case 'console':
       return <div id="console">TODO: Add a console view</div>;
     case 'performance':
+      if (eventWithSpans) {
+        eventWithSpans.entries[0].data = eventWithSpans?.entries[0]?.data?.filter(
+          datum => !datum?.data?.memory
+        );
+      }
```
Would love some feedback on this bit here. I couldn't figure out a nice way to filter out memory events so I memoized them prior to removing them from the data being passed to EventEntry. It feels gross so I opted to get eyes on it.

I'm not opposed to filtering within the EventEntry component but I wasn't sure if modifying another component for the sake of an experimental feature was kosher. Though to me that feels like the cleaner way to handle this.

billyvg replied:
Hmm I would probably handle this in `useReplayEvent` where we create the merged span in the first place. Thoughts on this @ryan953 ?

ryan953 replied:
I agree that `useReplayEvent` seems like the place to put this, shouldn't need useMemo or anything in this case. 

It is data that's derived from the state, and the hook will only update when state changes. So you can filter everything right at the `return` statement and it shouldn't run more often than it needs to.

We probably should have the hook return data that's pre-filtered for each use case, so each component gets exactly what it needs. Maybe the exported names would endup something like this:
```
return {
  fetchError: state.fetchError,
  fetching: state.fetching,
  onRetry,
  event,

  // ideally:
  chapterBreadcrumbs,
  performanceSpans,
  memorySpans,
};
```

Danondso replied:
I dig it, I'll make it happen.

Danondso replied:
Made a type for the memory spans called MemorySpanType and adjusted code to reflect that throughout the implementation. Hook now surfaces the memorySpans

billyvg commented on static/app/views/replays/detail/memoryChart.tsx:
```diff
@@ -0,0 +1,95 @@
+      seriesName: t('Used Heap Memory'),
+      data: memorySpans.map((span, i) => ({
+        value: span.data.memory.usedJSHeapSize,
+        name: i,
```
Hmm its been awhile since I've used the charts, we could do some pairing tomorrow though?

When you say they aren't playing right, what do you mean? Is it an issue with the units? e.g. seconds vs ms?

Danondso replied:
Oh! I bet it's because the timestamps are huge numbers, so they probably are stuck like way off the chart if I had to guess.

Danondso replied:
Nope I was wrong, I was accessing it via span.data.timestamp when it should have been span.timestamp. Pushing a fix once I format the times to something that isn't a bagillion seconds.

billyvg commented on static/app/views/replays/detail/memoryChart.tsx:
```diff
@@ -0,0 +1,95 @@
+  border: 1px solid ${p => p.theme.gray200};
+`;
+
+const NoMemoryMessageWrapper = styled('div')`
+  font-weight: 600;
+  font-size: ${p => p.theme.fontSizeLarge};
+  padding: ${space(2)};
+`;
```
I think we have an `<EmptyMessage>` component somewhere

Danondso replied:
Got it, next push will include the change.

Danondso replied:
Updated.

billyvg commented on static/app/views/replays/detail/memoryChart.tsx:
```diff
@@ -0,0 +1,95 @@
+const MemoryChartWrapper = styled('div')`
+  margin-top: -${space(0.75)};
+  border-radius: ${space(0.5)};
+  border: 1px solid ${p => p.theme.gray200};
```
I think there's a border alias we use

Danondso replied:
Will be updated in the next push.

ryan953 commented on static/app/views/replays/detail/memoryChart.tsx:
```diff
@@ -0,0 +1,95 @@
+        color: theme.gray100,
+      },
+      min: 0,
+      max: memorySpans.slice(-1)?.data?.memory.totalJSHeapSize,
```
What happens if the `max` isn't set? 

Rounding up to the nearest 10's should make the scale on the screen nicer, better to see `60MiB` instead of `57.2MiB` in my opinion. 
Maybe this is better as a follow up PR, where we set the scale based on the magnitude of the max value, so sometimes round to the nearest 10's, but maybe 50's or 100 if more memory is used

Danondso replied:
I'm actually not setting it, it seems like echarts interpolates based on the data input, so the grid looks the same.

Danondso replied:
I'll add it to this ticket. https://github.com/getsentry/sentry/issues/34110

ryan953 commented on static/app/views/replays/detail/memoryChart.tsx:
```diff
@@ -0,0 +1,95 @@
+      right: '10px',
+      top: '40px',
+      bottom: '10px',
+    },
```
what does this `grid` setting do? 

I would've guessed gridlines on the chart, horizontal or vertical, but i'm not seeing any of those... and left/right/top/bottom don't correlate to that idea either. So i'm lost.

ryan953 commented on static/app/views/replays/detail/memoryChart.tsx:
```diff
@@ -0,0 +1,95 @@
+  ) : (
+    <NoMemoryMessageWrapper>
+      {t('No memory metrics exist for replay.')}
+    </NoMemoryMessageWrapper>
```
Could we do an early `return <NoMemoryMessageWrapper>` if there are no memory spans? 

I find that type of change will make it easier in the future to find and update the error case and happy path independatly, instead of dealing with the extra ternary here. Having that ternary like we do now makes the extra object splat `{...chartOptions}` seem more necessary that it might be

Danondso replied:
Done

ryan953 commented on static/app/views/replays/detail/memoryChart.tsx:
```diff
@@ -0,0 +1,95 @@
+import {formatBytesBase2} from 'sentry/utils';
+import theme from 'sentry/utils/theme';
+
+function MemoryChart({memorySpans}) {
```
eslint fix after accepting this one, but we should type our props:

```suggestion
import type {RawSpanType} from 'sentry/components/events/interfaces/spans/types';

type Props = {
  memorySpans: RawSpanType[]
};

function MemoryChart({memorySpans}: Props) {
```

Danondso committed:
style: used the EmptyMessage component
fix: added timestamps in seconds as name to the memory chart and fixed the bounds.

Danondso committed:
feat: added new type MemorySpanType
refactor: made memory spans a first class prop on the useReplayEvent state
refactor: adjusted memoryGraph.tsx to accept props of MemorySpanType and resolve TS errors coming from that.

Danondso committed:
refactor: added memorySpanType as a prop and propagated that throughout the component
fix: slice(-1) was returning undefined and echarts was helping to correct that, so max for yAxis was removed.

Danondso committed:
feat: added timestamps to the graph!

Danondso committed:
refactor: return early if there's no memorySpans to work with.

Danondso committed:
Merge branch 'master' into feat/replays/33623/heap-memory-chart

billyvg commented on static/app/views/replays/detail/memoryChart.tsx:
```diff
@@ -0,0 +1,105 @@
+      top: '40px',
+      bottom: '10px',
+    },
+    tooltip: {
+      trigger: 'axis',
+      valueFormatter: (value: number | null) => formatBytesBase2(value || 0),
+    },
```
Ah I forgot to mention this earlier but we have a bunch chart components that setup defaults for all of these chartOptions. Could we try using them when possible? they're in `app/components/charts/components`. (let's do this as a follow-up PR)

Danondso replied:
Will post a ticket in this comment when I make it.

Danondso committed:
Merge branch 'master' into feat/replays/33623/heap-memory-chart

Danondso commented on static/app/views/replays/detail/memoryChart.tsx:
```diff
@@ -0,0 +1,105 @@
+    },
+  ];
+  return (
+    <MemoryChartContainer>
+      <MemoryChartWrapper>
```
I used the second one to add some space under the chart because the chart and the footer were right next to each other, let me give it another shot to see if I can eliminate it.

Danondso commented on static/app/views/replays/detail/memoryChart.tsx:
```diff
@@ -0,0 +1,105 @@
+        name: formatTimestamp(span.timestamp),
+      })),
+      stack: 'heap-memory',
+      color: theme.purple300,
```
Cool I'll not that in the ticket I make for the follow up work.

Danondso replied:
https://github.com/getsentry/sentry/issues/34110

Danondso committed:
style: added margin-bottom to the memoryChartWrapper component and eliminated the extra container component.

billyvg reviewed (approved):
:shipit:
